File: Pletor-canvas/backend/src/middlewares/errorHandler.ts
import type { FastifyError, FastifyReply, FastifyRequest } from 'fastify'
import { ZodError } from 'zod'

export class AppError extends Error {
  constructor(
    public statusCode: number,
    public code: string,
    message: string,
    public details?: unknown,
  ) {
    super(message)
    this.name = 'AppError'
  }
}

export class NotFoundError extends AppError {
  constructor(resource: string) {
    super(404, 'NOT_FOUND', `${resource} nebyl nalezen`)
  }
}

export function errorHandler(
  error: FastifyError,
  _request: FastifyRequest,
  reply: FastifyReply,
) {
  if (error instanceof AppError) {
    return reply.status(error.statusCode).send({
      error: error.code,
      message: error.message,
      details: error.details,
    })
  }

  if (error instanceof ZodError) {
    return reply.status(400).send({
      error: 'VALIDATION_ERROR',
      message: 'Neplatná vstupní data',
      details: error.errors.map((e) => ({
        field: e.path.join('.'),
        issue: e.message,
      })),
    })
  }

  // Neočekávaná chyba
  console.error('Neočekávaná chyba:', error)
  return reply.status(500).send({
    error: 'INTERNAL_ERROR',
    message: 'Interní chyba serveru',
  })
}

File: Pletor-canvas/backend-python/app/middlewares/error_handler.py
from fastapi import FastAPI, Request
from fastapi.exceptions import RequestValidationError
from fastapi.responses import JSONResponse


class AppError(Exception):
    def __init__(self, status_code: int, code: str, message: str, details: list[dict[str, str]] | None = None):
        self.status_code = status_code
        self.code = code
        self.message = message
        self.details = details or []
        super().__init__(message)


class NotFoundError(AppError):
    def __init__(self, entity: str):
        super().__init__(404, "NOT_FOUND", f"{entity} nebyl nalezen")


def register_error_handlers(app: FastAPI) -> None:
    @app.exception_handler(AppError)
    async def app_error_handler(_request: Request, exc: AppError) -> JSONResponse:
        body: dict[str, object] = {"error": exc.code, "message": exc.message}
        if exc.details:
            body["details"] = exc.details
        return JSONResponse(status_code=exc.status_code, content=body)

    @app.exception_handler(RequestValidationError)
    async def validation_error_handler(_request: Request, exc: RequestValidationError) -> JSONResponse:
        details = [{"field": ".".join(str(loc) for loc in e["loc"]), "issue": e["msg"]} for e in exc.errors()]
        return JSONResponse(
            status_code=422,
            content={"error": "VALIDATION_ERROR", "message": "Neplatný vstup", "details": details},
        )
